### PR TITLE
Correctly respect accessibilityActivationPoint

### PR DESCRIPTION
**Pull request by fabb**
Tries to solve #545

⚠️ Not yet finished, still need to add a test case for this + upside-down integration tests are failing.

### CODE REVIEW

File: WebDriverAgentLib/Categories/XCUIElement+FBTap.m [hunk 1]
```diff
@@ -22,26 +22,47 @@ @implementation XCUIElement (FBTap)
 - (BOOL)fb_tapWithError:(NSError **)error
 {
   NSValue *hitpointValue = self.lastSnapshot.suggestedHitpoints.firstObject;
-  CGPoint hitPoint = hitpointValue ? hitpointValue.CGPointValue : [self coordinateWithNormalizedOffset:CGVectorMake(0.5, 0.5)].screenPoint;
-  hitPoint.x -= self.frame.origin.x;
-  hitPoint.y -= self.frame.origin.y;
-  return [self fb_tapCoordinate:hitPoint error:error];
+  if (hitpointValue != nil) {
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBTap.m]: fast return would look better here:

```objc
if (nil == hitpointValue) {
  CGPoint hitPoint = [self hitPointCoordinate].screenPoint;
  return [self fb_tapScreenCoordinate:hitPoint error:error];
}
CGPoint hitPoint = hitpointValue.CGPointValue;
...
```

File: WebDriverAgentLib/Categories/XCUIElement+FBTap.m [hunk 2]
```diff
@@ -22,26 +22,47 @@ @implementation XCUIElement (FBTap)
...
+  } else {
+    // hitPointCoordinate might not be the center of the element frame, it depends on the accessibilityActivationPoint
+    CGPoint hitPoint = [self hitPointCoordinate].screenPoint;
+    return [self fb_tapScreenCoordinate:hitPoint error:error];
+  }
 }
 
 - (BOOL)fb_tapCoordinate:(CGPoint)relativeCoordinate error:(NSError **)error
 {
+  return [self fb_tapWithCoordinateBlock:^CGPoint {
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBTap.m]: is there any reason to wrap tap coordinates into a block?

File: WebDriverAgentLib/Commands/FBElementCommands.m [hunk 3]
```diff
@@ -327,15 +327,25 @@ + (NSArray *)routes
...
   if (nil == element) {
+    NSParameterAssert(xNumber != nil && yNumber != nil);
     XCUICoordinate *tapCoordinate = [self.class gestureCoordinateWithCoordinate:tapPoint application:request.session.application shouldApplyOrientationWorkaround:YES];
     [tapCoordinate tap];
   } else {
-    NSError *error;
-    if (![element fb_tapCoordinate:tapPoint error:&error]) {
-      return FBResponseWithError(error);
+    if (xNumber != nil && yNumber != nil) {
+      NSError *error;
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBElementCommands.m]: error variable can be defined once

File: WebDriverAgentLib/Commands/FBElementCommands.m [hunk 4]
```diff
@@ -327,15 +327,25 @@ + (NSArray *)routes
 + (id<FBResponsePayload>)handleTap:(FBRouteRequest *)request
 {
   FBElementCache *elementCache = request.session.elementCache;
-  CGPoint tapPoint = CGPointMake((CGFloat)[request.arguments[@"x"] doubleValue], (CGFloat)[request.arguments[@"y"] doubleValue]);
+  NSNumber *xNumber = request.arguments[@"x"];
+  NSNumber *yNumber = request.arguments[@"y"];
+  CGPoint tapPoint = CGPointMake((CGFloat)[xNumber doubleValue], (CGFloat)[yNumber doubleValue]);
   XCUIElement *element = [elementCache elementForUUID:request.parameters[@"uuid"]];
   if (nil == element) {
+    NSParameterAssert(xNumber != nil && yNumber != nil);
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBElementCommands.m]: I'd rather return explicit error response here (check _FBResponseWithErrorFormat_)

File: WebDriverAgentLib/Commands/FBElementCommands.m [hunk 5]
```diff
@@ -327,15 +327,25 @@ + (NSArray *)routes
...
     XCUICoordinate *tapCoordinate = [self.class gestureCoordinateWithCoordinate:tapPoint application:request.session.application shouldApplyOrientationWorkaround:YES];
     [tapCoordinate tap];
   } else {
-    NSError *error;
-    if (![element fb_tapCoordinate:tapPoint error:&error]) {
-      return FBResponseWithError(error);
+    if (xNumber != nil && yNumber != nil) {
+      NSError *error;
+      if (![element fb_tapCoordinate:tapPoint error:&error]) {
+        return FBResponseWithError(error);
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Commands/FBElementCommands.m]: this can be done once

File: WebDriverAgentLib/Categories/XCUIElement+FBTap.m [hunk 6]
```diff
@@ -22,26 +22,47 @@ @implementation XCUIElement (FBTap)
...
-  hitPoint.y -= self.frame.origin.y;
-  return [self fb_tapCoordinate:hitPoint error:error];
+  if (hitpointValue != nil) {
+    CGPoint hitPoint = hitpointValue.CGPointValue;
+    hitPoint.x -= self.frame.origin.x;
+    hitPoint.y -= self.frame.origin.y;
+    return [self fb_tapCoordinate:hitPoint error:error];
+  } else {
+    // hitPointCoordinate might not be the center of the element frame, it depends on the accessibilityActivationPoint
+    CGPoint hitPoint = [self hitPointCoordinate].screenPoint;
```
Comment (mykola-mokhnach) [inline WebDriverAgentLib/Categories/XCUIElement+FBTap.m]: have you tested if this works properly in landscape?

Comment (fabb) [inline WebDriverAgentLib/Categories/XCUIElement+FBTap.m]: i have executed the unit tests in `FBTapTest`, the landscape tests work fine. as written above, the upside-down tests yet fail though.

File: WebDriverAgentLib/Categories/XCUIElement+FBTap.m (hunk 2 above)
Comment (fabb) [inline WebDriverAgentLib/Categories/XCUIElement+FBTap.m]: yes, `fb_waitUntilFrameIsStable` should be executed first, i didn't want to change the execution order compared to the original source code

File: WebDriverAgentLib/Commands/FBElementCommands.m (hunk 5 above)
Comment (fabb) [inline WebDriverAgentLib/Commands/FBElementCommands.m]: imho not really nicer to read, but ok: 415ebe8663

File: WebDriverAgentLib/Commands/FBElementCommands.m (hunk 3 above)
Comment (fabb) [inline WebDriverAgentLib/Commands/FBElementCommands.m]: 415ebe8663

File: WebDriverAgentLib/Categories/XCUIElement+FBTap.m (hunk 1 above)
Comment (fabb) [inline WebDriverAgentLib/Categories/XCUIElement+FBTap.m]: f449d29be1

File: WebDriverAgentLib/Commands/FBElementCommands.m (hunk 4 above)
Comment (fabb) [inline WebDriverAgentLib/Commands/FBElementCommands.m]: df6b6716f2

**Comment by fabb** [comment]
@mykola-mokhnach any idea why the upside-down integration tests are failing? (`testTapCoordinatesInPortraitUpsideDown` and `testTapInPortraitUpsideDown`)

**Comment by mykola-mokhnach** [comment]
@fabb Because of XCTest bugs :(
Try to apply coordinates transformations similar to FBInvertPointForApplication to workaround the problem

**Comment by fabb** [comment]
@mykola-mokhnach Yeah I read the comment about iOS 10, but when using `hitPointCoordinate` it seems to work correctly for landscape left/right, strange that only upside-down is an issue...

**Comment by mykola-mokhnach** [comment]
What can I say... Welcome to the club

**Comment by marekcirkos** [comment]
Hey, thank you for your work on this. Few notes here:
1) How do you know that `accessibilityActivationPoint` is returned by `screenCoordinates`?
2) Please add tests for case you are fixing.

**Comment by fabb** [comment]
@marekcirkos 
1. `accessibilityActivationPoint` is in screen coordinates, as described in Apple Documentation: "The activation point for the accessibility element, in screen coordinates.". `hitPointCoordinate` unfortunately has no documentation, as ist is a private method from XCTest - it uses `accessibilityActivationPoint` internally though, as sherlocked via breakpoints.
2. Sure, just need some more time for this

**Comment by marekcirkos** [comment]
Is it finished?

**Comment by fabb** [comment]
No, didn't get to it yet. Help appreciated, pretty stuffed at the moment. Need to add a test case + fix esiting failing tests.

**Comment by marekcirkos** [comment]
Got diff coming that should solve your problem

**Comment by fabb** [comment]
Awesome, can you link it here when it's done?

**Comment by marekcirkos** [comment]
97144d0e6ed108d01e1d7d9ebb0c489a0dabbafd

**Comment by marekcirkos** [comment]
and stack around it :)

**Comment by fabb** [comment]
Nice! What about the `handleTap` method in `FBElementCommands`? Shouldn't it call `tapWithError` in case no x/y coordinate parameters are provided?

**Comment by marekcirkos** [comment]
`handleTap` handles case when you specify where to tap screen or element anyway. If you want just to activate element you should use click endpoint. 

@mykola-mokhnach Does `Appium` depend on this endpoint? I was thinking about splitting it in to two use cases. Move tap element on x,y to `click` and leave tap screen on x,y.

**Comment by mykola-mokhnach** [comment]
@marekcirkos Yes we have some dependencies, but they can be easily updated if necessary. Just let us know when the change is merged to master.
@imurchie FYI

**Pull request closed without merging**